File: Kernel Simulator/Misc/Calculation/Calc.vb

'    Kernel Simulator  Copyright (C) 2018-2021  EoflaOE
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Public Module Calc

    ''' <summary>
    ''' Try to calculate an expression
    ''' </summary>
    ''' <param name="Expression">An expression</param>
    ''' <returns>A dictionary of results and success/failure</returns>
    ''' <remarks>Obsolete in favor of string evaluator.</remarks>
    <Obsolete>
    Public Function DoCalc(ByVal Expression As String) As Dictionary(Of Double, Boolean)
        Dim Calculated As Boolean
        Dim Result As Double
        Dim DictCheck As New Dictionary(Of Double, Boolean) 'Key: Result / Value: If True, then success. Else, failure.
        Try
            Wdbg("I", "Expression: {0}", Expression)
            Result = New DataTable().Compute(Expression, Nothing)
            Wdbg("I", "Result: {0}", Result)
            Calculated = True
        Catch ex As Exception
            Wdbg("E", "Error while evaluating expression: {0}", ex.Message)
            WStkTrc(ex)
        End Try
        DictCheck.Add(Result, Calculated)
        Return DictCheck
    End Function

End Module

File: public/Nitrocid.Addons/Nitrocid.Extras.Calculators/Commands/Calc.cs
//
// Nitrocid KS  Copyright (C) 2018-2024  Aptivi
//
// This file is part of Nitrocid KS
//
// Nitrocid KS is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Nitrocid KS is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY, without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

using System;
using Nitrocid.ConsoleBase.Colors;
using Nitrocid.ConsoleBase.Writers;
using Terminaux.Writer.ConsoleWriters;
using Nitrocid.Kernel.Debugging;
using Nitrocid.Languages;
using Nitrocid.Shell.ShellBase.Commands;
using StringMath;

namespace Nitrocid.Extras.Calculators.Commands
{
    /// <summary>
    /// Calculates expressions.
    /// </summary>
    /// <remarks>
    /// This command lets you calculate expressions and return the results.
    /// </remarks>
    class CalcCommand : BaseCommand, ICommand
    {

        public override int Execute(CommandParameters parameters, ref string variableValue)
        {
            try
            {
                double Res = ((MathExpr)parameters.ArgumentsText).Result;
                DebugWriter.WriteDebug(DebugLevel.I, "Res = {0}", Res);
                TextWriterColor.Write(parameters.ArgumentsText + " = " + Res.ToString());
                variableValue = Res.ToString();
                return 0;
            }
            catch (Exception ex)
            {
                DebugWriter.WriteDebug(DebugLevel.I, "Error trying to calculate expression {0}: {1}", parameters.ArgumentsText, ex.Message);
                DebugWriter.WriteDebugStackTrace(ex);
                TextWriters.Write(Translate.DoTranslation("Error in calculation.") + " {0}", true, KernelColorType.Error, ex.Message);
                return ex.GetHashCode();
            }
        }

    }
}
